pkg/controller/kmeshsecurity/security_client.go

```diff
@@ -0,0 +1,199 @@
+		 opts:     opts,
+	 }
+
+	 conn, err := nets.GrpcConnect(CSRSignAddress);
+	 if err != nil {
+		 log.Errorf("Failed to connect to endpoint %s: %v", opts.CAEndpoint, err)
+		 return nil, fmt.Errorf("failed to connect to endpoint %s", opts.CAEndpoint)
+	 }
+ 
+	 c.conn = conn
+	 c.client = pb.NewIstioCertificateServiceClient(conn)
+	 return c, nil
+ }
+
+// CSRSign calls Citadel to sign a CSR.
+func (c *CitadelClient) CSRSign(csrPEM []byte, certValidTTLInSec int64) (res []string, err error) {
+	crMetaStruct := &structpb.Struct{
+		Fields: map[string]*structpb.Value{
+			security.CertSigner: {
+				Kind: &structpb.Value_StringValue{StringValue: c.opts.CertSigner},
+			},
+		},
+	}
+	req := &pb.IstioCertificateRequest{
+		Csr:              string(csrPEM),
+		ValidityDuration: certValidTTLInSec,
+		Metadata:         crMetaStruct,
+	}
+
+	defer func() {
+		if err != nil {
+			log.Errorf("failed to sign CSR: %v", err)
+			if err := c.reconnect(); err != nil {
+				log.Errorf("failed reconnect: %v", err)
+			}
+		}
+	}()
+
+	ctx := metadata.NewOutgoingContext(context.Background(), metadata.Pairs("ClusterID", c.opts.ClusterID))
+	resp, err := c.client.CreateCertificate(ctx, req)
+	if err != nil {
+		return nil, fmt.Errorf("create certificate: %v", err)
+	}
+
+	if len(resp.CertChain) <= 1 {
+		return nil, errors.New("invalid empty CertChain")
+	}
+
+	return resp.CertChain, nil
+}
+
+// concatCerts concatenates PEM certificates, making sure each one starts on a new line
+func concatCerts(certsPEM []string) []byte {
+	if len(certsPEM) == 0 {
+		return []byte{}
+	}
+	var certChain bytes.Buffer
+	for i, c := range certsPEM {
+		certChain.WriteString(c)
+		if i < len(certsPEM)-1 && !strings.HasSuffix(c, "\n") {
+			certChain.WriteString("\n")
+		}
+	}
+	return certChain.Bytes()
+}
+
+func (c *CitadelClient) fetch_cert(uid string) (secret *security.SecretItem, err error) {
+	var rootCertPEM []byte
+	
+	workloadCache := workload.GetCacheByUid(uid)
+	if workloadCache == nil {
+		log.Errorf("get workloadCache failed")
+		err := errors.New("get workloadCache failed")
+		return nil, err
+	}
+	csrHostName := &spiffe.Identity{
+		TrustDomain:    workloadCache.TrustDomain, 
+		Namespace:      workloadCache.Namespace, 
+		ServiceAccount: workloadCache.ServiceAccount,
+	}
+
+	options := pkiutil.CertOptions{
+		Host:       csrHostName.String(),
+		RSAKeySize: c.opts.WorkloadRSAKeySize,
+		PKCS8Key:   c.opts.Pkcs8Keys,
+		ECSigAlg:   pkiutil.SupportedECSignatureAlgorithms(c.opts.ECCSigAlg),
+		ECCCurve:   pkiutil.SupportedEllipticCurves(c.opts.ECCCurve),
+	}
+
+	// Generate the cert/key, send CSR to CA.
+	csrPEM, keyPEM, err := pkiutil.GenCSR(options)
+	if err != nil {
+		log.Errorf("%s failed to generate key and certificate for CSR: %v", workloadCache.Name, err)
+		return nil, err
+	}
+	certChainPEM, err := c.CSRSign(csrPEM, int64(c.opts.SecretTTL.Seconds()))
+	if err != nil {
+		err = nil
+		certChainPEM, err = c.CSRSign(csrPEM, int64(c.opts.SecretTTL.Seconds()))
```

Comment:
Try again immediately? Or do you need to retry at certain intervals?
And only once? Does it need to be done more than once?